File: frontend/src/api/analytics.ts
// src/api/analytics.ts
import { api } from '@/lib/api';

export type KeywordHit = { keyword: string; count: number };
export type HourlyPoint = { hour: string; count: number };
export type TalkgroupHit = { groupId: string; display?: string; count: number };

const isMock = import.meta.env.VITE_MOCK === '1';

const mockKeywordHits: KeywordHit[] = [
  { keyword: 'pursuit', count: 12 },
  { keyword: 'accident', count: 9 },
  { keyword: 'alarm', count: 7 },
];

const mockHourly: HourlyPoint[] = [
  { hour: '00', count: 4 }, { hour: '01', count: 3 }, { hour: '02', count: 2 },
  { hour: '03', count: 2 }, { hour: '04', count: 1 }, { hour: '05', count: 1 },
  { hour: '06', count: 2 }, { hour: '07', count: 5 }, { hour: '08', count: 7 },
  { hour: '09', count: 6 }, { hour: '10', count: 8 }, { hour: '11', count: 9 },
  { hour: '12', count: 11 }, { hour: '13', count: 12 }, { hour: '14', count: 10 },
  { hour: '15', count: 9 }, { hour: '16', count: 8 }, { hour: '17', count: 7 },
  { hour: '18', count: 10 }, { hour: '19', count: 12 }, { hour: '20', count: 14 },
  { hour: '21', count: 13 }, { hour: '22', count: 9 }, { hour: '23', count: 6 },
];

const mockTopTalkgroups: TalkgroupHit[] = [
  { groupId: '7017-38529', display: 'SAFD Alert', count: 42 },
  { groupId: '7017-38273', display: 'SAPD NORTH A PTL', count: 38 },
  { groupId: '7017-39043', display: 'Fire Tac 3', count: 22 },
];

export type DashboardMetrics = {
  total_calls_24h: number;
  active_playlists: number;
  transcripts_today: number;
  avg_transcription_confidence: number;
  processing_queue_size: number;
  api_calls_today: number;
  recent_calls: any[];
  top_talkgroups: TalkgroupHit[];
};

const mockDashboardMetrics: DashboardMetrics = {
  total_calls_24h: 145,
  active_playlists: 8,
  transcripts_today: 127,
  avg_transcription_confidence: 0.82,
  processing_queue_size: 3,
  api_calls_today: 247,
  recent_calls: [],
  top_talkgroups: mockTopTalkgroups
};

export async function fetchKeywordHits(): Promise<KeywordHit[]> {
  if (isMock) return mockKeywordHits;
  try {
    const res = await api.get<KeywordHit[]>('/analytics/keywords');
    return res.data ?? [];
  } catch (error) {
    console.error('Failed to fetch keyword hits', error);
    throw error;
  }
}

export async function getDashboardMetrics(): Promise<DashboardMetrics> {
  if (isMock) return mockDashboardMetrics;
  try {
    const res = await api.get<DashboardMetrics>('/analytics/dashboard');
    return res.data;
  } catch (error) {
    console.error('Failed to fetch dashboard metrics', error);
    throw error;
  }
}

export async function fetchHourlyActivity(): Promise<HourlyPoint[]> {
  if (isMock) return mockHourly;
  try {
    const res = await api.get<HourlyPoint[]>('/analytics/hourly');
    return res.data ?? [];
  } catch (error) {
    console.error('Failed to fetch hourly activity', error);
    throw error;
  }
}

export async function fetchTopTalkgroups(limit = 10): Promise<TalkgroupHit[]> {
  if (isMock) return mockTopTalkgroups.slice(0, limit);
  try {
    const res = await api.get<TalkgroupHit[]>(`/analytics/talkgroups/top?limit=${limit}`);
    const data = res.data ?? [];
    return data.slice(0, limit);
  } catch (error) {
    console.error('Failed to fetch top talkgroups', error);
    throw error;
  }
}

// Export aliases for Dashboard compatibility
export const getKeywordHits = fetchKeywordHits;

File: app_api/routers/analytics.py
from fastapi import APIRouter, Query, Depends
from typing import List, Dict, Any
from datetime import datetime, timedelta
import asyncpg

from database import get_pool
from models.analytics import (
    KeywordHit, HourlyPoint, TalkgroupHit, DashboardMetrics,
    QualityDistribution, ApiMetricsSummary
)
from models.calls import CallMetadata

router = APIRouter()


def transform_dashboard_metrics(metrics: DashboardMetrics) -> Dict[str, Any]:
    """
    Transform dashboard metrics to include frontend-expected field names.

    Adds camelCase aliases:
    - 'active_playlists' → 'feedCount', 'activeFeeds'
    - 'total_calls_24h' → 'recentCalls'
    - 'transcripts_today' → 'transcriptsToday'
    - Keeps original snake_case fields for backward compatibility
    """
    # Convert Pydantic model to dict
    result = metrics.model_dump() if hasattr(metrics, 'model_dump') else metrics.dict()

    # Add camelCase aliases
    result['feedCount'] = result.get('active_playlists', 0)
    result['activeFeeds'] = result.get('active_playlists', 0)
    result['recentCalls'] = result.get('total_calls_24h', 0)
    result['transcriptsToday'] = result.get('transcripts_today', 0)

    return result


@router.get("/dashboard")
async def get_dashboard_metrics(
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get comprehensive dashboard metrics."""
    async with pool.acquire() as conn:
        # Get total calls in last 24h
        total_calls = await conn.fetchval(
            "SELECT COUNT(*) FROM bcfy_calls_raw WHERE started_at > NOW() - INTERVAL '24 hours'"
        )

        # Get active playlists
        active_playlists = await conn.fetchval(
            "SELECT COUNT(*) FROM bcfy_playlists WHERE sync = TRUE"
        )

        # Get transcripts today
        transcripts_today = await conn.fetchval(
            "SELECT COUNT(*) FROM transcripts WHERE created_at >= CURRENT_DATE"
        )

        # Get avg transcription confidence
        avg_confidence = await conn.fetchval(
            "SELECT AVG(confidence) FROM transcripts WHERE created_at >= CURRENT_DATE"
        ) or 0.0

        # Get processing queue size
        queue_size = await conn.fetchval(
            "SELECT COUNT(*) FROM processing_state WHERE status = 'queued'"
        ) or 0

        # Get API calls today (with fallback if table doesn't exist)
        try:
            api_calls_today = await conn.fetchval(
                "SELECT COUNT(*) FROM api_call_metrics WHERE timestamp >= CURRENT_DATE"
            ) or 0
        except asyncpg.UndefinedTableError:
            api_calls_today = 0

        # Get recent calls
        recent_calls = await conn.fetch(
            """
            SELECT call_uid, feed_id, tg_id, started_at, duration_ms
            FROM bcfy_calls_raw
            ORDER BY started_at DESC
            LIMIT 5
            """
        )

        # Get top talkgroups
        top_talkgroups = await conn.fetch(
            """
            SELECT tg_id, COUNT(*) as count
            FROM bcfy_calls_raw
            WHERE started_at > NOW() - INTERVAL '24 hours' AND tg_id IS NOT NULL
            GROUP BY tg_id
            ORDER BY count DESC
            LIMIT 10
            """
        )

    metrics = DashboardMetrics(
        total_calls_24h=total_calls or 0,
        active_playlists=active_playlists or 0,
        transcripts_today=transcripts_today or 0,
        avg_transcription_confidence=float(avg_confidence),
        processing_queue_size=queue_size,
        api_calls_today=api_calls_today,
        recent_calls=[dict(row) for row in recent_calls],
        top_talkgroups=[
            TalkgroupHit(tg_id=row["tg_id"], count=row["count"])
            for row in top_talkgroups
        ]
    )

    return transform_dashboard_metrics(metrics)


@router.get("/hourly", response_model=List[HourlyPoint])
async def get_hourly_activity(
    hours: int = Query(24, ge=1, le=168),
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get hourly call activity."""
    query = """
        SELECT
            TO_CHAR(DATE_TRUNC('hour', started_at), 'HH24:00') as hour,
            COUNT(*) as count
        FROM bcfy_calls_raw
        WHERE started_at > NOW() - ($1::int * INTERVAL '1 hour')
        GROUP BY DATE_TRUNC('hour', started_at)
        ORDER BY DATE_TRUNC('hour', started_at)
    """

    async with pool.acquire() as conn:
        rows = await conn.fetch(query, hours)

    return [HourlyPoint(hour=row["hour"], count=row["count"]) for row in rows]


@router.get("/talkgroups/top", response_model=List[TalkgroupHit])
async def get_top_talkgroups(
    limit: int = Query(10, ge=1, le=100),
    period: str = Query("24h"),
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get top active talkgroups."""
    # Convert period to hours for parameterized query
    hours_map = {
        "1h": 1,
        "24h": 24,
        "7d": 168,
    }
    interval_hours = hours_map.get(period, 24)

    query = """
        SELECT tg_id, COUNT(*) as count
        FROM bcfy_calls_raw
        WHERE started_at > NOW() - ($1::int * INTERVAL '1 hour') AND tg_id IS NOT NULL
        GROUP BY tg_id
        ORDER BY count DESC
        LIMIT $2
    """

    async with pool.acquire() as conn:
        rows = await conn.fetch(query, interval_hours, limit)

    return [TalkgroupHit(tg_id=row["tg_id"], count=row["count"]) for row in rows]


@router.get("/keywords", response_model=List[KeywordHit])
async def get_keyword_hits(
    limit: int = Query(10, ge=1, le=100),
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get keyword hit counts (placeholder - needs keywords table)."""
    # TODO: Implement when keywords table/system is available
    return []


@router.get("/transcription-quality", response_model=QualityDistribution)
async def get_transcription_quality(
    pool: asyncpg.Pool = Depends(get_pool)
):
    """Get transcription quality distribution."""
    query = """
        SELECT
            SUM(CASE WHEN confidence > 0.85 THEN 1 ELSE 0 END) as excellent,
            SUM(CASE WHEN confidence > 0.7 AND confidence <= 0.85 THEN 1 ELSE 0 END) as good,
            SUM(CASE WHEN confidence > 0.5 AND confidence <= 0.7 THEN 1 ELSE 0 END) as fair,
            SUM(CASE WHEN confidence <= 0.5 THEN 1 ELSE 0 END) as poor
        FROM transcripts
        WHERE created_at >= CURRENT_DATE
    """

    async with pool.acquire() as conn:
        row = await conn.fetchrow(query)

    return QualityDistribution(
        excellent=row["excellent"] or 0,
        good=row["good"] or 0,
        fair=row["fair"] or 0,
        poor=row["poor"] or 0
    )
